File: android/reactnativeview/src/test/java/com/lmig/uscm/digital/appplatform/reactnativeview/BiometricModuleTests.kt
package com.lmig.uscm.digital.appplatform.reactnativeview

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import kotlin.test.assertEquals

/**
 * BiometricModule Tests
 */
class BiometricModuleTests {

    private var biometricModule: BiometricModule? = null

    @Before
    fun setUp() {
        val mockReactAppContext = Mockito.mock(ReactApplicationContext::class.java)
        biometricModule = BiometricModule(reactContext = mockReactAppContext)
    }

    @After
    fun tearDown() {
    }

    @Test
    fun testGetName() {
        assertEquals("Biometric", biometricModule?.name)
    }

    @Test
    fun testIsSupported() {
        class TestPromise: Promise {
            override fun resolve(value: Any?) {
                val result = value as Boolean
                result?.let{
                    // Should be false as we are running against a mock Context
                    assertEquals(false, it)
                }
            }
            override fun reject(code: String?, message: String?) {}
            override fun reject(code: String?, e: Throwable?) {}
            override fun reject(code: String?, message: String?, e: Throwable?) {}
            override fun reject(message: String?) {}
            override fun reject(reason: Throwable?) {}

        }
        val testProm = TestPromise()
        biometricModule?.isSupported(promise = testProm)
    }

    @Test
    fun testGetBiometricType() {
        class TestPromise: Promise {
            override fun resolve(value: Any?) {
                val result = value as String
                result?.let{
                    // Should be false as we are running against a mock Context
                    assertEquals("fingerprint", it)
                }
            }
            override fun reject(code: String?, message: String?) {}
            override fun reject(code: String?, e: Throwable?) {}
            override fun reject(code: String?, message: String?, e: Throwable?) {}
            override fun reject(message: String?) {}
            override fun reject(reason: Throwable?) {}

        }
        val testProm = TestPromise()
        biometricModule?.getBiometricType(promise = testProm)
    }

    // Has Credentials, Get Credentials, Save Credentials & Delete Credentials cannot be easily tested as they
    // require interaction with SharedPreferences
}

File: ios/BiometricModuleTests.swift
//
//  BiometricModuleTests.swift
//  AppPlatformTests
//
//  Created by McKim, Mark (LIT) on 13/11/2017.
//  Copyright © 2017 Liberty Mutual. All rights reserved.
//

import XCTest
@testable import AppPlatform

class BiometricModuleTests: XCTestCase {
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    func testIsSupporte() {
        let biometric = BiometricModule()
        XCTAssert(biometric.isSupported() == false)
    }
    func testGetBiometricType() {
        let biometric = BiometricModule()
        XCTAssert(biometric.getBiometricType() == "none")
    }
    // Has Credentials, Get Credentials, Save Credentials & Delete Credentials cannot be easily tested as they
    // require interaction with User Defaults
}
